nibabel/streamlines/compact_list.py

```diff
@@ -0,0 +1,232 @@
+import numpy as np
+
+
+class CompactList(object):
+    """ Class for compacting list of ndarrays with matching shape except for
+    the first dimension.
+    """
+
+    BUFFER_SIZE = 87382*4  # About 4 Mb if item shape is 3 (e.g. 3D points).
+
+    def __init__(self, iterable=None):
+        """
+        Parameters
+        ----------
+        iterable : iterable (optional)
+            If specified, create a ``CompactList`` object initialized from
+            iterable's items. Otherwise, create an empty ``CompactList``.
+
+        Notes
+        -----
+        If `iterable` is a ``CompactList`` object, a view is returned and no
+        memory is allocated. For an actual copy use the `.copy()` method.
+        """
+        # Create new empty `CompactList` object.
+        self._data = np.array(0)
+        self._offsets = np.array([], dtype=int)
+        self._lengths = np.array([], dtype=int)
+
+        if isinstance(iterable, CompactList):
+            # Create a view.
+            self._data = iterable._data
+            self._offsets = iterable._offsets
+            self._lengths = iterable._lengths
+
+        elif iterable is not None:
+            offsets = []
+            lengths = []
+            # Initialize the `CompactList` object from iterable's item.
+            offset = 0
+            for i, e in enumerate(iterable):
+                e = np.asarray(e)
+                if i == 0:
+                    new_shape = (CompactList.BUFFER_SIZE,) + e.shape[1:]
+                    self._data = np.empty(new_shape, dtype=e.dtype)
+
+                end = offset + len(e)
+                if end >= len(self._data):
+                    # Resize needed, adding `len(e)` new items plus some buffer.
+                    nb_points = len(self._data)
+                    nb_points += len(e) + CompactList.BUFFER_SIZE
+                    self._data.resize((nb_points,) + self.shape)
+
+                offsets.append(offset)
+                lengths.append(len(e))
+                self._data[offset:offset+len(e)] = e
+                offset += len(e)
+
+            self._offsets = np.asarray(offsets)
+            self._lengths = np.asarray(lengths)
+
+            # Clear unused memory.
+            if self._data.ndim != 0:
+                self._data.resize((offset,) + self.shape)
+
+    @property
+    def shape(self):
```

Comment:
This is a slightly puzzling use of `shape` isn't it?   For example if you passed in 20 streamlines, of shapes (n_0, 3), (n_1, 3) ... (n_19, 3), then `shape == 3`?  Could you call this something else like `common_shape`?